content/blog/2022-04-07-compiling-black-with-mypyc-part-2.md

```diff
@@ -0,0 +1,537 @@
+
+```console
+$ python -m cProfile -o profile.pstats -m black src/black/__init__.py --check
+All done! ✨ 🍰 ✨
+1 file would be left unchanged.
+
+$ gprof2dot profile.pstats | dot -Tsvg -o profile.svg
+```
+
+With the help of [gprof2dot] (well, actually the [yelp-gprof2dot] fork), I converted the
+profiling data into a nice SVG graph which I could then open in my web browser.
+
+> Tip: please open this massive SVG in a new tab because viewing it here will probably be
+> painful :)
+
+![gprof2dot graph showing where time was spent in a call tree](/media/cProfileGraph-black-1.svg)
+
+I tried using [Scalene] as I've heard good things about it, but it didn't work sadly. It
+wasn't that bad as I still had [py-spy], [line_profiler], and good ol' cProfile. py-spy in
+particular was invaluable since it can profile (well err sample) C extensions which
+cProfile cannot. line_profiler was used exclusively for micro-opimitizations :p
+
+Anyway, I repeated this process quite a few times, making sure to try different files to
+get a general feel where time is going regardless of the input. Here are the main
+takeaways:
+
+- Initial parsing with blib2to3 takes up 30-50% of formatting runtime!
+
+- The AST equivalent safety check is cheap at 5-10% (obviously only when changes were
+  made, otherwise it's zero 🙂)
+
+- The actual formatting logic's runtime is mostly spent in the CST visitor usually taking
+  up 3/4. The rest went to the transformers which handle line breaks, string literals, and
+  some special cases.
+
+I was quite surprised how much time blib2to3 related functions were eating up. Just look
+how concentrated the hotspots really are!
+
+![gprof2dot graph showing where time was spent parsing](/media/cProfileGraph-black-blib2to3-1.svg)
+
+These hotspots make this part of the codebase way easier to optimize, so I optimized
+blib2to3 first.
+
+It's been so long since I first looked into this, so I don't remember what optimizations I
+tried initially, but I do remember them having no effect :(
+
+I tried other things and they actually helped 🎉 ... probably since I took into account how
+mypyc works. Ultimately, many different optimizations were done over three rounds.
+
+#### Tightening up existing type annotations
+
+The stricter your type annotations are in your codebase, the more invariants mypyc will be
+able to infer. It'll then use this information and write type-specific code that is
+faster. This code won't work if it gets an object of a different type, but that's why we
+use and enforce type annotations, so we (and mypyc) can safely assume it's going to be the
+right type.
+
+This involves reading through the code and control flow, checking whether certain states
+are impossible. Blib2to3 is mostly a legacy codebase, so we when added type annotations to
+it, it was done to unblock other work. The goal was to make blib2to3 type check, and not
+write the best typed code ever. So, naturally there were a few permissive (parameter) type
```

Comment:
Does this make the code run faster because mypyc can represent `Text` variables more efficiently than `Optional[Text]`?

Reply:
In theory yes as it shouldn't have to do type checks for None values, but the examples I show have it narrowed by the asserts eitherway so I need to look into all of this eitherway.

Reply:
It does reduce the function call overhead so it's less of a benefit than I previously thought, but it still helps which means I don't have to rewrite this part.